app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoUI.kt

```diff
@@ -0,0 +1,249 @@
+                                    OutlinedButton(
+                                            onClick = { onResetPuzzle() },
+                                            modifier = Modifier.weight(1f)) {
+                                        Icon(imageVector = Icons.Filled.Refresh,
+                                            tint = MaterialTheme.colors.onSurface,
+                                            modifier = Modifier.size(16.dp),
+                                            contentDescription = null)
+                                        Text(text = "RETRY",
+                                            color = MaterialTheme.colors.onSurface,
+                                            modifier = Modifier.padding(start = 8.dp))
+                                    }
+                                }
+                                if (state.continueButtonVisible) {
+                                    if(hasNextPuzzle) {
+                                        Button(onClick = { onNextPuzzle() },
+                                            modifier = Modifier.weight(1f)) {
+                                            Text(text = "CONTINUE")
+                                        }
+                                    } else {
+                                        Button(onClick = { onBack() },
+                                            modifier = Modifier.weight(1f)) {
+                                            Text(text = "DONE")
+                                        }
+                                    }
+                                }
+                            }
+                            if (state.continueButtonVisible || (state.rating?.rating ?: 0) > 0) {
+                                Row(modifier = Modifier.height(32.dp)
+                                            .align(Alignment.CenterHorizontally)
+                                        ) {
+                                    RatingBar(
+                                        rating = state.rating?.rating?.toFloat() ?: 0f,
+                                        onTap = { onRate(it) },
+                                        modifier = Modifier
+                                            .align(Alignment.CenterVertically)
+                                    )
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        } ?: run {
+            Box(
+                contentAlignment = Alignment.Center,
+                modifier = Modifier.fillMaxSize()
+                    .padding(horizontal = 8.dp, vertical = 4.dp),
+            ) {
+                Text(
+                    text = "Loading...",
+                    color = MaterialTheme.colors.onBackground,
+                )
+            }
+        }
+    }
+    LaunchedEffect(solved) {
+        snapshotFlow { solved }
```

Comment:
This looks like logic that needs to be offloaded to the ViewModel. Not sure I even understand it, is that a way of saying "if it's solved, call onSolve()"?

Reply:
yes, pretty much. It should be in the VM, you're right